### PR TITLE
Add new metrics to monitor memory

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

To introduce new metrics that will be used to monitor the memory usage as requested by security.

# Changes

* Introduce `ii_archive_stable_memory_bytes`, `ii_archive_heap_pages` and `ii_archive_heap_memory_bytes` metrics to the archive canister.
* Introduce `internet_identity_heap_memory_bytes` and `internet_identity_stable_memory_bytes` metrics to main II canister.

### CODE REVIEW

File: src/archive/src/main.rs [hunk 1]
```diff
@@ -575,6 +575,7 @@ fn http_request(req: HttpRequest) -> HttpResponse {
 }
 
 fn encode_metrics(w: &mut MetricsEncoder<Vec<u8>>) -> std::io::Result<()> {
+    const WASM_PAGE_SIZE_IN_BYTES: f64 = 65536.0;
```
Comment (frederikrothenberger) [inline src/archive/src/main.rs:578]: There should be a constant for that in the `cdk`. Also, as a side-note: I would not do floating point calculations unless absolutely necessary. Floating point calculations could introduce rounding errors.

File: src/archive/src/main.rs [hunk 2]
```diff
@@ -634,6 +635,23 @@ fn encode_metrics(w: &mut MetricsEncoder<Vec<u8>>) -> std::io::Result<()> {
         stable_size() as f64,
         "Number of stable memory pages used by this canister.",
     )?;
+    w.encode_gauge(
+        "ii_archive_stable_memory_bytes",
```
Comment (frederikrothenberger) [inline src/archive/src/main.rs:639]: Is this useful? Do you need both pages and bytes?

**Review by frederikrothenberger — COMMENTED** [review]
Honestly, I would just go with `ii_archive_heap_pages` which is clearly missing. The rest can be deduced from `pages` when processing the data.

File: src/archive/src/main.rs (hunk 2 above)
Comment (lmuntaner) [reply src/archive/src/main.rs:639]: Yes, there was the agreement to expose the bytes and alarms will be created for us.

See the thread: https://dfinity.slack.com/archives/C063Z1ZBL8P/p1728300749385249?thread_ts=1727341770.248009&cid=C063Z1ZBL8P

File: src/archive/src/main.rs (hunk 1 above)
Comment (lmuntaner) [reply src/archive/src/main.rs]: Nice! Fixed!

**Comment by lmuntaner** [comment]
> Honestly, I would just go with `ii_archive_heap_pages` which is clearly missing. The rest can be deduced from `pages` when processing the data.

Alarms for memory threshold will be added if we expose this metrics.

File: src/archive/src/main.rs (hunk 2 above)
Comment (frederikrothenberger) [reply src/archive/src/main.rs:638]: Ah, TIL. That makes sense 👍

**Review by frederikrothenberger — APPROVED** [review]
LGTM, thanks!

**Pull request merged**